spanner/api/Spanner/CancelBackupOperation.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Cloud.Spanner.Admin.Database.V1;
+using Google.Cloud.Spanner.Common.V1;
+using Google.LongRunning;
+using Google.Protobuf.WellKnownTypes;
+using log4net;
+using System;
+using static GoogleCloudSamples.Spanner.Program;
+
+namespace GoogleCloudSamples.Spanner
+{
+    public class CancelBackupOperation
+    {
+        static readonly ILog s_logger = LogManager.GetLogger(typeof(CancelBackupOperation));
+
+        // [START spanner_cancel_backup_create]
+        public static object SpannerCancelBackupOperation(
+            string projectId, string instanceId, string databaseId, string backupId)
+        {
+            // Create the DatabaseAdminClient instance.
+            DatabaseAdminClient databaseAdminClient = DatabaseAdminClient.Create();
+
+            // Initialize request parameters.
+            Backup backup = new Backup
+            {
+                Database = DatabaseName.Format(projectId, instanceId, databaseId),
+                ExpireTime = DateTime.UtcNow.AddDays(14).ToTimestamp()
+            };
+            string parent = InstanceName.Format(projectId, instanceId);
+
+            // Make the CreateBackup request.
+            Operation<Backup, CreateBackupMetadata> response =
+                databaseAdminClient.CreateBackup(parent, backup, backupId);
+
+            // Create the OperationsClient instance and execute CancelOperation.
+            OperationsClient operationsClient = OperationsClient.Create();
+            operationsClient.CancelOperation(response.Name);
+
+            s_logger.Info("Waiting for the operation to finish.");
+
+            // Poll until the long-running operation is complete. It will
+            // either complete or be cancelled.
+            Operation<Backup, CreateBackupMetadata> completedResponse =
+                response.PollUntilCompleted();
+
+            if (!completedResponse.IsFaulted)
+            {
```

Comment:
Hmm I think it would be a little more idiomatic to poll inside a try catch:
... = response.PollUntilCompleted().ThrowOnAnyError();

And actually, is CancelOperation long running? Would it failed if the backup was already created? Because if so, I think that the clearest thing would be to poll and throw CancelOperation inside the try catch instead.

Reply:
Yes `CancelOperation` is long running but it never throws any exception except `google.rpc.Code.UNIMPLEMENTED` when API is not implemented.
 
It has to check by `GetOperation` and if the value of status code is `Code.CANCELLED` then CancelOperation has been successfully cancel, but we never know when this CancelOperation will Finish and we don't have any Poll option on same.

Reply:
Just to clarify, I don't think the `CancelOperation` is long-running. It sends the cancel request to the server which registers the intent and returns immediately. The backend then goes about cancelling the `CreateBackup` operation in its own time. It's possible for the backup to be created before the backend tries to cancel it. Hence the `PollUntilCompleted()` above.

@laljikanjareeya is right about checking for the `Code.CANCELLED` in the original `CreateBackup` operation to check if it was actually cancelled before the backup could be completed.

Reply:
OK, I understand this is the APIs design, it's just I find it weird to have to poll the operation being cancelled and check if it failed to actually know that it was cancelled.
Ths intuitive thing for me would have been to poll the cancel operation itself and if that fails then it's because the backup was created before you could cancel it.
I suppose it is too late to try to influence that?

Reply:
Yeah, I agree that would've been more intuitive. Unfortunately it's too late to change that behavior in the backend since the APIs were finalized sometime earlier last year.